### PR TITLE
Handle `enabled=off` with recursion in place

### PR DESCRIPTION
**Pull request by jimklimov**
(Includes and) Follows up from #623 with a practical feature: if the same intermediate dataset that declares `enabled=off` also says `recursive=on`, the disablement takes place for descendants as well.

As discussed among ideas in #623, this PR adds recognition for datasets (descendants of a "backupSet") which declare not only an `org.znapzend:enabled` value but optionally also an `org.znapzend:recursive` one (so only these one or two ZFS properties are defined in a local source dataset under the one which defines a full retention schedule).

It also somewhat reasonably defines behavior for such dataset trees where a `backupSet` with a full retention schedule has some descendant which defines an `enabled=off` behavior (individual or recursive), and some of its descendants re-enable the backup (individually or recursively). I'm sure this can be tested with mocks, but out of time at least today, maybe longer, to pursue testing. Worked as expected in a real-life setup (which however did not try all edge cases) :)

This is also seen in `znapzendzetup list` outputs:
````
...
*** backup plan: rpool/home ***
           dst_0 = znapzend:pond/export/DUMP/NUTCI/znapzend/ci-deb/rpool/home
      dst_0_plan = 1week=>1day,1month=>1week,1year=>1month,10years=>6months
         enabled = on
         mbuffer = off
    mbuffer_size = 128M
   post_znap_cmd = off
    pre_znap_cmd = off
       recursive = on
             src = rpool/home
        src_plan = 1week=>1day,1month=>1week,1year=>1month,10years=>6months
        tsformat = znapzend-auto-%Y-%m-%dT%H:%M:%SZ
      zend_delay = 0

*** backup plan: rpool/home/abuild/.ccache ***
         enabled = off
             src = rpool/home/abuild/.ccache

*** backup plan: rpool/home/abuild/jenkins-nut-altroots ***
         enabled = off
       recursive = on
             src = rpool/home/abuild/jenkins-nut-altroots

...
````

### CODE REVIEW

File: lib/ZnapZend.pm [hunk 1]
```diff
@@ -145,6 +145,216 @@ my $killThemAll = sub {
...
+                my $propIR;
+
+                # if the property does not exist, the command will just return.
+                # In this case, use the default determined above.
+                $propLE = <$propLE>; # || $enabled_default;
+                if ($propLE) {
+                    chomp($propLE);
+                }
+
+                $propLR = <$propLR>;
```
Comment (oetiker) [inline lib/ZnapZend.pm:202]: same here

File: lib/ZnapZend.pm [hunk 2]
```diff
@@ -145,6 +145,216 @@ my $killThemAll = sub {
...
+                                    }
+                                }
+                            }
+                            elsif ($nearestLE eq "on" && defined($propIR) && $propIR eq "on") {
+                                # zfs-inheritance of "enabled=on"
+                                $isDisabled = -1;
+                            }
+                        }   # else only "enabled=whatever" is defined in some ancestor, but
+                            # not "recursive" anywhere => this one remains enabled, probably
+                    }   # else no ancestor defines a local "enabled" setting
```
Comment (oetiker) [inline lib/ZnapZend.pm:334]: you are making use of && a lot here instead of the looser `and` intentional ?

File: CHANGES [hunk 3]
```diff
@@ -3,6 +3,7 @@ znapzend (0.21.3) unstable; urgency=medium
   * Maintenance release: refine splitting of [[user@]host:]dataset[:with-colons][@snap[:with-colons]] strings to work for the realistic majority of use-cases; fix back support of pool root dataset in such spec
   * Update self-tests with verification that [[user@]host:]dataset[:with-colons][@snap[:with-colons]] string decoding yields expected results
   * Fixed CI recipes and contents for spell-checker
+  * Extended handling of "enabled=off" for sub-trees: if the same intermediate dataset declares "recursive=on", the disablement takes place for descendants as well
```
Comment (oetiker) [inline CHANGES:6]: maybe explain the idea in a few more words ... also updating the documentation on enabled=off might be helpful for people to understand what is going to happen

File: lib/ZnapZend.pm [hunk 4]
```diff
@@ -145,6 +145,216 @@ my $killThemAll = sub {
...
+                open my $propLE, '-|', @cmdLE;
+
+                my @cmdLR = (@{$self->zZfs->priv}, qw(zfs get -H -s local -o value org.znapzend:recursive), $dataSet);
+                print STDERR '# ' . join(' ', @cmdLR) . "\n" if $self->debug;
+                open my $propLR, '-|', @cmdLR;
+                my $propIR;
+
+                # if the property does not exist, the command will just return.
+                # In this case, use the default determined above.
+                $propLE = <$propLE>; # || $enabled_default;
```
[the remark this answers is not in the document]
Comment (jimklimov) [reply lib/ZnapZend.pm:197]: Copy-pasta from the original code (see the removed lines originally in `createSnapshot`) :)

File: lib/ZnapZend.pm (hunk 2 above)
Comment (jimklimov) [reply lib/ZnapZend.pm:334]: Rather, forgot the difference - hopping too many languages lately :) Do you suggest I change these?

File: CHANGES (hunk 3 above)
Comment (jimklimov) [reply CHANGES:6]: Good point :)

**Comment by jimklimov** [comment]
Thanks for the review, and the master-branch bump (docker fix), nice to see it all claim to be green again :)

> removing the extra snapshots quickly seems like a good move!

I don't think this PR tackles it directly - this is part of `createSnapshot()` and was for a while; this change here (and started by "externalization" of code in #623) modifies how users can state more easily which datasets are not-enabled for znapzend handling so their snapshots would be removed earlier. Also #626 takes care of skipping `zfs send` for whatever datasets are in the not-enabled lists (also builds upon #623, but is independent of the change in this PR).

File: lib/ZnapZend.pm (hunk 2 above)
Comment (oetiker) [reply lib/ZnapZend.pm:334]: well I am normally weary of using && for its high binding power

File: lib/ZnapZend.pm (hunk 4 above)
[the remark this answers is not in the document]
Comment (oetiker) [reply lib/ZnapZend.pm:197]: hihi that does not make it better :)

**Comment by jimklimov** [comment]
Hm, I don't actually see any docs about the `enabled=off` handling :\

File: lib/ZnapZend.pm (hunk 4 above)
[the remark this answers is not in the document]
Comment (jimklimov) [reply lib/ZnapZend.pm]: PR update posted

File: lib/ZnapZend.pm [hunk 5]
```diff
@@ -145,6 +145,216 @@ my $killThemAll = sub {
...
+
+                $propLR = <$propLR>;
+                if ($propLR) {
+                    chomp($propLR);
+                } else {
+                    my @cmdIR = (@{$self->zZfs->priv}, qw(zfs get -H -s inherited -o value org.znapzend:recursive), $dataSet);
+                    print STDERR '# ' . join(' ', @cmdIR) . "\n" if $self->debug;
+                    open $propIR, '-|', @cmdIR;
+
+                    $propIR = <$propIR>;
```
[the remark this answers is not in the document]
Comment (jimklimov) [reply lib/ZnapZend.pm]: PR update posted

File: lib/ZnapZend.pm (hunk 2 above)
Comment (jimklimov) [reply lib/ZnapZend.pm]: PR update posted

**Comment by jimklimov** [comment]
Ok, so after the merge of groundwork from #623 the specific changes of this PR are much better visible :)

**Review by oetiker — APPROVED** [review]

**Pull request merged**